Make `restic mount` faster and consume less memory

aawsome opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------

In the FUSE implementation restic data structure have been read (meaning loading blobs from the repository) and internal data structures have been created when a FUSE "node" was created. This was the case, for instance, for all item within a dir when this dir was accessed. This lead to quite some memory usage and made the FUSE filesystem "unresponsive". 

This PR changes this behavior and only reads restic data structure when the information is really needed. Moreover, the internal data structure have been optimized. Also concurrent operation within FUSE should now work correctly. (FUSE documentation says that access should be synced which wasn't)

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
closes #1680 

Thanks @greatroar for issuing #2585 which makes me think about whether the bottleneck really is only reading tree blobs.

Checklist
---------

- [X] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- I have not added tests for all changes in this PR
- I have not added documentation for the changes (in the manual)
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [X] I have run `gofmt` on the code in all commits
- [X] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [X] I'm done, this Pull Request is ready for review

greatroar commented on internal/fuse/file.go:
```diff
@@ -111,6 +90,27 @@ func (f *file) Read(ctx context.Context, req *fuse.ReadRequest, resp *fuse.ReadR
 	debug.Log("Read(%v, %v, %v), file size %v", f.node.Name, req.Size, req.Offset, f.node.Size)
 	offset := req.Offset
 
+	if !f.hasSizes {
```
Shouldn't this be protected by a lock? Or does FUSE handle that?

greatroar replied:
Never mind. FUSE calls Lookup, and Lookup returns a fresh *file every time. So this is safe.

aawsome replied:
Thanks for researching! I wasn't sure about how FUSE handles concurrency, but `getBlobAt` also seems not to be concurrency-safe, so I assumed it would be handled by FUSE...

greatroar replied:
Actually, I failed to spot that the bazil.org/fuse package stores the node it gets from Lookup in a cache, and it's concurrent. So there really should be locks all over internal/fuse.

greatroar replied:
False alarm. It looks like the kernel driver will synchronize access, as long as you don't set the AsyncRead MountOption. I've empirically verified this by setting minSnapshotsReloadTime to 0, running restic mount with the race detector on, reading the snapshots directory concurrently in two loops and adding snapshots.

MichaelEischer replied:
The documentation of bazil/fuse actually says that synchronization is required (see https://godoc.org/bazil.org/fuse#hdr-Service_Methods):

> Multiple goroutines may call service methods simultaneously; the methods being called are responsible for appropriate synchronization.

I'd assume that some caching in the kernel effectively synchronizes access, but it probably not a good idea to rely on it.

aawsome replied:
I added a Mutex to synchronize access

MichaelEischer commented:
It might be an option to implement the `NodeOpener` which is called when a file is opened. That way the chunk size collection would happen immediately when opening a file and is not delayed until reading the first bytes.

The main difference is the point in time when the user would be notified in case some blobs are missing. The current implementation in restic probably fails with an error before providing any details on the file node, with the change this is delayed until the first bytes are read. I think that the new semantic is actually what a user would expect: Reading a damaged file fails, but usually you are able to see the metadata.

aawsome commented:
@MichaelEischer Thanks for searching the manual and offering the NodeOpener option which I'm using now.
I tried to do performance tests of this new version but did not really succeed to see differences when comparing this to. In fact it seems that there a lot of variation w.r.t. performance when using FUSE (maybe some kernel caching?)...
Is there anyone who can help out with testing this change?

Memory footprint is however clearly reduced.

ProactiveServices commented:
@aawsome Happy to help, do you have any specific tests you'd like to perform? I have several hundred-gig repos I can test with.

aawsome commented:
@ProactiveServices Thank you very much! I'm especially interested in response times and CPU usage and comparison between "plain" restic and a version including this PR.
The code changes should apply for listing and reading files, e.g. `ls file` or `cat file`/`cp file somewhere`.

I used `/usr/bin/time restic mount` to get CPU timings and a simple `time` before the commands, e.g. `time ls file` to get response times.

aawsome commented:
With the last commit, also dir operations should be faster (and a bit less memory consuming)...

greatroar commented on internal/fuse/file.go:
```diff
@@ -20,47 +21,31 @@ import (
 const blockSize = 512
 
 // Statically ensure that *file implements the given interface
```
This comment is now stale because one of the checks is for *openFile. You could update it, but removing it may be easier.

aawsome replied:
the comment is now updated.

greatroar commented on internal/fuse/file.go:
```diff
@@ -136,9 +138,20 @@ func (f *file) Read(ctx context.Context, req *fuse.ReadRequest, resp *fuse.ReadR
+	//
+	// Hence Lock to ensure operation on blobs is not done concurrently
+	// (doesn't work with cached blob in getBlobAt)
+	f.m.Lock()
```
defer f.m.Unlock()? There are multiple exits, and lack of memory can cause a panic.

aawsome replied:
is changed.

aawsome committed:
Optimize FUSE - make command `restic mount` faster and consume less memory

- Add Open() functionality to dir
- only access index for blobs when file is read
- Implement NodeOpener and put one-time file stuff there
- Add comment about locking as suggested by bazil.org/fuse

=> Thanks at Michael Eischer for suggesting the last two improvements

aawsome commented:
rebased to master and added the changes suggested by @greatroar

aawsome commented:
Rebased this PR after #2790 has been merged.

aawsome commented:
Noticed that the new blob cache by @greatroar already implements locking; hence I removed the locks from `file.go` and just left a comment there.

aawsome commented:
rebased after #2787 has been merged.

MichaelEischer commented on internal/fuse/dir.go:
```diff
@@ -20,10 +21,12 @@ var _ = fs.NodeStringLookuper(&dir{})
 	inode       uint64
 	parentInode uint64
 	node        *restic.Node
+	isOpen      bool
```
`isOpen` seems redundant. Afaict `isOpen` == `items != nil`

aawsome replied:
Ah - thanks for the hint! Changed it.

MichaelEischer commented on internal/fuse/dir.go:
```diff
@@ -72,35 +65,48 @@ func replaceSpecialNodes(ctx context.Context, repo restic.Repository, node *rest
+	}, nil
+}
+
+func (d *dir) Open() error {
```
Is this `Open` method used by any external caller or just internally by ReadDirAll and Lookup? In the latter case please change the signature to `func (d *dir) open(ctx context.Context) error`. That avoid also the problematic use of `context.TODO()`

aawsome replied:
changed it to `open(context.Context)`

MichaelEischer commented on internal/fuse/dir.go:
```diff
@@ -133,6 +139,11 @@ func (d *dir) calcNumberOfLinks() uint32 {
+	if err != nil {
+		return nil, err
+	}
+	debug.Log("called")
```
I think we can remove that `debug.Log` call

aawsome replied:
yes - this was just to debug; is removed.

MichaelEischer commented on internal/fuse/dir.go:
```diff
@@ -170,6 +181,10 @@ func (d *dir) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
+	if err != nil {
+		return nil, err
+	}
 	debug.Log("Lookup(%v)", name)
```
please move the `debug.Log` call above `d.Open()`

aawsome replied:
is changed.

aawsome commented on changelog/unreleased/issue-1680:
```diff
@@ -0,0 +1,7 @@
+`restic mount` is now more responsive and uses less memory.
+
+https://github.com/restic/restic/issues/1680
+https://github.com/restic/restic/pull/2587
```
is added.

aawsome commented on internal/fuse/dir.go:
```diff
@@ -20,10 +21,12 @@ var _ = fs.NodeStringLookuper(&dir{})
 	node        *restic.Node
+	isOpen      bool
+	m           sync.Mutex
+	items       map[string]*restic.Node
```
changed.

MichaelEischer reviewed (approved):
LGTM. I've force-pushed the branch to retrigger the CI. Seems like the flaky rclone test error has changed into a flaky rclone test crash, after #2855.
